src/test.ts

```diff
@@ -130,12 +130,20 @@ export function testTree(tree: Tree, expect: string, mayIgnore = defaultIgnore)
     throw new Error(`Unexpected end of tree. Expected ${stack[0].slice(pos[0]).map(s => s.name).join(", ")} at ${tree.length}\n${tree}`)
 }
 
+function toLineContext(file: string, index: number) {
+  const endIndex = Math.min(index + 80, file.length);
+
+  return file.substring(index, endIndex).split(/\n/).map(str => '  | ' + str).join('\n');
+}
+
```

Comment:
Why the `|` markers? I think showing a slice like this, without even taking care to round up/down to line boundaries, produces rather crude output. Wouldn't just outputting the line number be more useful?

Reply:
Well, the output looks exactly as shown in the PR description which I personally find helpful.

Showing line numbers unfortunately won't work, as you (or really anyone) can fiddle around with the file contents before passing it to the function. We [do it here, for instance](https://github.com/lezer-parser/lezer-generator/blob/master/test/test-cases.ts#L28) to remove the grammar to be parsed.

As for the rounding, it is not necessary. The regular expression will always match line boundaries so the start index always refers to the start of a line.

Reply:
Good point about the line numbers and the start of line. If you add some logic that scans forward to the next line end when there is one or the end of the string otherwise this looks good.

Reply:
https://github.com/lezer-parser/lezer-generator/pull/4/commits/be4aa4e860eedb31dae6c580eedd7d2e7f0b0834 ensures we print full 80 chars + end of last line.